### PR TITLE
Issue256 add typesafe tests

### PR DESCRIPTION
**Pull request by cecol**
for #256 
I add `NullAndOptionalConfig`, `ListExample`, and `TypesafeConfigErrorsExample`.
Because the `TypesafeConfigMapSpec` looks like `MapExample` and for avoiding duplicated tests, I am not sure that whether add it or not.

### CODE REVIEW

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 1]
```diff
@@ -0,0 +1,405 @@
+package zio.config.typesafe
+
+import com.typesafe.config.ConfigRenderOptions
+import zio.config._
+import zio.config.magnolia.DeriveConfigDescriptor.descriptor
+import zio.test.Assertion._
+import zio.test.{ DefaultRunnableSpec, suite, test, _ }
+
+object ListSpec extends DefaultRunnableSpec {
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:9]: I think I have already done this in separate test. So you could remove `test("A complex enough case for testing HOCON List config and write to/back of HOCON source")`

Comment (afsalthaj) [reply typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:9]: Have a look at `TypesafeConfigReadWriteTest` . Its covering read-write of complex structures

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+      val readWritten = written match {
+        case Left(v)          => Left(v)
+        case Right(configSource) => read(descriptor[A] from configSource)
+      }
+
+      assert(readWritten)(equalTo(zioConfigResult)) &&
+      assert(zioConfigResult)(equalTo(expectedReadResult))
+    },
+    test("A kebab case for testing HOCON List config") {
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:225]: This one is ok

File: typesafe/src/test/scala/zio/config/typesafe/NullAndOptionalConfigSpec.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+)
+
+object EmployeeDetails {
+
+  /**
+   * An example, where manual configuration program is much more richer, with more details involved in it.
+   * This is one such situation where corresponding automatic description won't work. The automatic description
+   * will be just based on keys, and you can ofcourse manipulate keys later on but there is times when you really
+   * need to describe your little configurations
+   */
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/NullAndOptionalConfigSpec.scala:26]: Remove these comments

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+      // Any list which you think could be empty in your application configuration file, which actually shouldn't be,
+      // You can choose to have `Option[NonEmptyList]` to avoid the error outcome
+      // Have a read through https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
+      final case class Extra2(
+        ci: String,
+        vi: Either[Int, Either[Long, Either[Double, Either[Float, String]]]],
+        lst: List[Int],
+        vvv: Option[List[Int]]
+      )
+      final case class Extra(hi: String, bi: String, r: List[Extra2])
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:347]: Rename `Extra` to `Details2`

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+      final case class Details(
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:348]: Rename `Details` to `Details1`

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+          |  port {
+          |    va : ba
+          |  }
+          |}
+          |""".stripMargin
+
+      // Any list which you think could be empty in your application configuration file, which actually shouldn't be,
+      // You can choose to have `Option[NonEmptyList]` to avoid the error outcome
+      // Have a read through https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
+      final case class Extra2(
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:341]: Rename `Extra2` to `Details2`

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+          |database {
+          |  port {
+          |    va : ba
+          |  }
+          |}
+          |""".stripMargin
+
+      // Any list which you think could be empty in your application configuration file, which actually shouldn't be,
+      // You can choose to have `Option[NonEmptyList]` to avoid the error outcome
+      // Have a read through https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
```
Comment (afsalthaj) [inline typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:340]: Remove comments in tests

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala (hunk 1 above)
Comment (cecol) [reply typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala:8]: updated

File: typesafe/src/test/scala/zio/config/typesafe/NullAndOptionalConfigSpec.scala (hunk 3 above)
Comment (cecol) [reply typesafe/src/test/scala/zio/config/typesafe/NullAndOptionalConfigSpec.scala]: updated

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala (hunk 5 above)
Comment (cecol) [reply typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala]: updated

File: typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala (hunk 7 above)
Comment (cecol) [reply typesafe/src/test/scala/zio/config/typesafe/ListSpec.scala]: updated

**Comment by afsalthaj** [comment]
@cecol I tried to update the branch through GitHub UI. I think if you could merge with the latest changes, we are good to go

**Comment by cecol** [comment]
updated

> @cecol I tried to update the branch through GitHub UI. I think if you could merge with the latest changes, we are good to go

I merged the latest master

**Pull request merged**